router/pkg/pubsub/nats.go

```diff
@@ -43,46 +57,93 @@ func (p *natsPubSub) ID() string {
 
 func (p *natsPubSub) ensureConn() error {
 	if p.conn == nil {
-		return errors.New("NATS is not configured")
+		return newEDFSNatsError(errors.New("NATS is not configured"))
 	}
 	return nil
 }
 
-func (p *natsPubSub) Subscribe(ctx context.Context, topic string, updater resolve.SubscriptionUpdater) error {
+func (p *natsPubSub) Subscribe(ctx context.Context, subjects []string, updater resolve.SubscriptionUpdater, streamConfiguration *pubsub_datasource.StreamConfiguration) error {
 	if err := p.ensureConn(); err != nil {
-		return err
+		return newEDFSNatsError(fmt.Errorf(`failed to ensure nats connection: %w`, err))
 	}
-	sub, err := p.conn.SubscribeSync(topic)
-	if err != nil {
-		return fmt.Errorf("error subscribing to NATS topic %s: %w", topic, err)
+	if streamConfiguration != nil {
+		js, err := jetstream.New(p.conn)
+		if err != nil {
+			return newEDFSNatsError(fmt.Errorf(`failed to create jetstream: %w`, err))
+		}
+
+		consumer, err := js.CreateOrUpdateConsumer(ctx, streamConfiguration.StreamName, jetstream.ConsumerConfig{
+			Durable:        streamConfiguration.Consumer, // Durable consumers are not removed automatically regardless of the InactiveThreshold
+			FilterSubjects: subjects,
+		})
+		if err != nil {
+			return newEDFSNatsError(fmt.Errorf(`failed to create or update consumer "%s": %w`, streamConfiguration.Consumer, err))
+		}
+		if consumer == nil {
+			return newEDFSNatsError(fmt.Errorf(`consumer "%s" is nil; it is likely the nats stream "%s" does not exist`, streamConfiguration.Consumer, streamConfiguration.StreamName))
+		}
+		go func() {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+				msgBatch, err := consumer.FetchNoWait(1)
+				if err != nil {
+					return
+				}
+				for msg := range msgBatch.Messages() {
+					err = msg.Ack()
+					if err != nil {
+						return
+					}
+					updater.Update(msg.Data())
+				}
+			}
+		}()
+		return nil
+	}
+
+	msgChan := make(chan *nats.Msg)
+	subscriptions := make([]*nats.Subscription, len(subjects))
+	for _, subject := range subjects {
+		subscription, err := p.conn.ChanSubscribe(subject, msgChan)
+		if err != nil {
+			return newEDFSNatsError(fmt.Errorf(`error subscribing to NATS subject "%s": %w`, subject, err))
+		}
+		subscriptions = append(subscriptions, subscription)
```

Comment:
It's a nitpick, but you could also add and index "i" in line 110 and assign using the index, no need to append.

Reply:
Got it